### PR TITLE
BaseHeader counts lines with only spaces in header_start processing

### PR DESCRIPTION
**Pull request by taldcroft**
`BaseHeader.process_lines()` doesn't throw away lines that have only one or more spaces, while `BaseData.process_lines()` does.  This is inconsistent and super confusing if you have a file that requires specifying `header_start` and `data_start`.

The fix is pretty trivial, but I'm worried it constitutes an API change since people may have code that is working in the context of the current inconsistent behavior.  So we should defer this past 0.4 and think a bit about it.

### CODE REVIEW

**Comment by taldcroft** [comment]
@hamogu - PR attached.

Closes #3816.  (Duplicate)

**Comment by taldcroft** [comment]
@mdmueller - can you have a look at the change to `tokenizer.c` to be sure it is correct?

File: docs/io/ascii/read.rst [hunk 1]
```diff
@@ -199,7 +244,7 @@ be the string ``'0'``,
 otherwise you may get unexpected behavior [#f1]_.  By default the
 ``<missing_spec>`` is applied to all columns unless column name strings are
 supplied.  An alterate way to limit the columns is via the
-``fill_include_names`` and ``fill_exclude_names`` keyword arguments in |read|.
+n``fill_include_names`` and ``fill_exclude_names`` keyword arguments in |read|.
```
Comment (taldcroft) [inline docs/io/ascii/read.rst]: Argh, editor mishap.  Will fix.

**Comment by mdmueller** [comment]
The tokenizer changes look fine to me. :+1:

File: astropy/io/ascii/tests/test_read.py [hunk 2]
```diff
@@ -992,3 +992,51 @@ def test_probably_html():
...
+               'skip this line',  # line 2
+               '1 2'],  # line 3
+              [{'header_start': 1, 'data_start': 3}]),
+
+             (['# comment',
+               '',
+               ' ',
+               'a b',  # line 0
+               '',
+               ' ',
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_read.py]: I think we say "white-space character" not just "space" in the docs. So maybe one example should have a "\t".

File: docs/io/ascii/read.rst [hunk 3]
```diff
@@ -140,6 +138,53 @@ Parameters for ``read()``
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The three parameters ``header_start``, ``data_start`` and ``data_end`` make it
+possible to read a table file that has extraneous non-table data included.
+This is a case where you need to help out ``io.ascii`` and tell it where to
+find the header and data.
+
+When processing of a file into a header and data components any blank lines
+(which might have whitespace characters) and commented lines (starting with the
+comment character ``#``) are stripped out *before* the header and data
```
Comment (hamogu) [inline docs/io/ascii/read.rst]: comment character, *typically* ``#`` ...
We have several reader with different comment characters e.g. LaTex (Ok, that's ost complex anyway), daophot (OK, also more complex). Still, comment character is a parameter we can set.

**Comment by hamogu** [comment]
Apart fomr the two minor in-line comments I think this looks good.

**Comment by hamogu** [comment]
Not that I want to complain, but I thought if you add a test (for non-whitespace) then commonly people would would wait for Travis to pass before merging instead of [skip ci]?

Anyway, thanks for fixing this and let's hope it does not break too much stuff in users code that relied on the broken behavior.

**Comment by taldcroft** [comment]
@hamogu - I ran tests locally on py2.7 and 3.4 and decided not to burn travis / appveyor time.  All the tests get run on the new commits to master and I get an email if there is a failure.  So yes, bending the rules slightly and taking a small chance.

**Comment by taldcroft** [comment]
And fortunately it did pass: https://travis-ci.org/astropy/astropy/builds/65870940.

**Comment by astrofrog** [comment]
We should remember to check the Windows builds too

**Comment by taldcroft** [comment]
I saw the windows builds passing on the astropy status page but didn't mention it.
